File: exercise/scala/day22/src/main/scala/eid/SerialNumber.scala
package eid

opaque type SerialNumber = Int

object SerialNumber {
  private val parsingErrorMessage = "Serial number should be between 1 and 999"

  def parse(input: String): Either[ParsingError, SerialNumber] =
    input.toEitherParsing(x => isValid(x), parsingErrorMessage)

  def apply(value: Int): SerialNumber = {
    require(isValid(value), parsingErrorMessage)
    value
  }

  private def isValid(value: Int): Boolean = value >= 1 && value <= 999
}

File: exercise/php/day22/src/Eid/SerialNumber.php
<?php

namespace Eid;

use InvalidArgumentException;
use Monad\Option;
use Monad\Option\None;
use Monad\Option\Some;

class SerialNumber
{
    private int $value;

    private function __construct(int $value)
    {
        if (!self::isValid($value)) {
            throw new InvalidArgumentException("Serial number should be between 1 and 999");
        }
        $this->value = $value;
    }

    private static function isValid(int $value): bool
    {
        return $value >= 1 && $value <= 999;
    }

    public static function parse(string $input): Option
    {
        $value = IntFunctions::toInt($input);
        if ($value !== null && self::isValid($value)) {
            return Some::create(new SerialNumber($value));
        }
        return new None(new ParsingError("Year should be positive and less than 100"));
    }

    public function __toString()
    {
        return sprintf("%03d", $this->value);
    }
}
